### DISCUSSION TITLE
Tracing with OpenTelemetry

### DISCUSSION THREAD

**Discussion by astrojuanlu** [kedro-org/kedro#3858] [Idea]
> Quick 2 cents
> 
> - +1 on setting the default level _of the Kedro logger_ to `INFO` (if just for backwards compatibility), and then having `-q` set it to `WARNING` and `-qq` to `ERROR`
>   - I don't think the current `logging.yml` logic is the only way to achieve that though
>   - I'm also ambivalent on whether we should change the _global_ logging level to `INFO`
> - -1 on keeping the current `logging.yml` logic - whoever wants fine grained control of logs, file logging, rotation etc should be using journald, supervisor, Datadog, or whatever other solution. this is _not_ Kedro's responsibility
> - +1 on directing users to current best-practices for logging, including creating official integrations and docs for Sentry, OpenTelemetry

_Originally posted by @astrojuanlu in https://github.com/kedro-org/kedro/issues/3446#issuecomment-1978476734_

Inspired by: https://github.com/mlflow/mlflow/pull/11454/

**Comment by astrojuanlu** [comment]
This is possible and not very hard:

![image](https://github.com/kedro-org/kedro/assets/316517/7fa4a282-945a-471f-a286-5116284f0f7c)

However, there are some loose ends: https://github.com/kedro-org/kedro/issues/3978

**Comment by astrojuanlu** [comment]
Yesterday a MLE told me that basically everyone expects a production system to have observability, and that it makes no sense that we don't have a native otel integration by now.

**Reply by noklam** [reply]
What does a native otel integration mean? Does that mean a custom Kedro otel instrumentation, TraceProvider? What exactly is preventing someone to do this already?

**Reply by lariel-fernandes** [reply]
IMO a reasonable otel integration would be: Having an official Kedro hook that starts/ends spans for each pipeline, node, maybe even for each data catalog load/save operation. That would already make pipeline observability way easier out of box.

There's nothing preventing anyone from doing it with custom hooks though.

**Comment by astrojuanlu** [comment]
Another platform supporting traces: https://www.traceloop.com/

```python
from traceloop.sdk import Traceloop

Traceloop.init()

...

from traceloop.sdk.decorators import workflow

@workflow(name="suggest_answers")
def suggest_answers(question: str):
  ...

```

![image](https://github.com/user-attachments/assets/aeef6f58-28ef-44ff-a6ec-be494608d680)

**Comment by astrojuanlu** [comment]
Another platform with OpenTelemetry integration https://betterstack.com/docs/logs/open-telemetry/

![image](https://github.com/user-attachments/assets/fbd9da31-e54a-44df-8386-6bf80d190b0e)

(image from https://betterstack.com/community/guides/logging/json-logging/)

**Comment by astrojuanlu** [comment]
Another platform _built_ on OpenTelemetry https://signoz.io/docs/

![image](https://github.com/user-attachments/assets/d632b011-6f75-49cb-b31d-b9102f42c0cf)

(image from https://signoz.io/blog/json-logs/)
